File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Call/Views/ParticipantsView/ParticipantsView.m
//
//  ParticipantsView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 27.09.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "ParticipantsView.h"
#import "ParticipantVideoView.h"
#import "ParticipantView.h"
#import "CallParticipant.h"
#import "LocalVideoView.h"
#import "Profile.h"

@interface ParticipantsView()
//MARK: - IBOutlets
@property (weak, nonatomic) IBOutlet UIStackView *topStackView;
@property (weak, nonatomic) IBOutlet UIStackView *bottomStackView;
//MARK: - Properties
@property (strong, nonatomic) CallInfo *callInfo;
@property (assign, nonatomic) QBRTCConferenceType conferenceType;
@end

@implementation ParticipantsView
//MARK: - Public Methods
- (void)setupWithCallInfo:(CallInfo *)callInfo conferenceType:(QBRTCConferenceType)conferenceType {
    self.conferenceType = conferenceType;
    self.callInfo = callInfo;
    
    __weak __typeof(self)weakSelf = self;
    [self.callInfo setOnUpdatedParticipant:^(NSNumber * _Nonnull userID, NSString * _Nonnull fullName) {
        __typeof(weakSelf)strongSelf = weakSelf;
        ParticipantView *participantView = (ParticipantView *)[strongSelf participantViewWithUserID:userID];
        participantView.name = fullName;
    }];
    
    for (CallParticipant *participant in self.callInfo.interlocutors) {
        UIView *participantView = self.conferenceType == QBRTCConferenceTypeVideo ?
        [self createParticipantVideoViewWithParticipant:participant]
        : [self createParticipantViewWithParticipant:participant] ;
        BOOL viewIsFull = self.topStackView.arrangedSubviews.count == 2;
        if (viewIsFull) {
            [self.bottomStackView addArrangedSubview:participantView];
            return;
        }
        [self.topStackView addArrangedSubview:participantView];
    }
}

- (void)addLocalVideo:(UIView *)videoView {
    ParticipantVideoView *participantView = (ParticipantVideoView *)[self participantViewWithUserID:self.callInfo.localParticipantId];
    if (!participantView) {
        CallParticipant *participant = [self.callInfo participantWithId:self.callInfo.localParticipantId];
        participantView = (ParticipantVideoView *)[self createParticipantVideoViewWithParticipant:participant];
        [self.bottomStackView addArrangedSubview:participantView];
    }
    participantView.videoView = videoView;
}

- (void)setupVideoTrack:(QBRTCVideoTrack *)videoTrack participantId:(NSNumber *)participantId {
    ParticipantVideoView *participantView = (ParticipantVideoView *)[self participantViewWithUserID:participantId];
    if (participantView) {
        QBRTCRemoteVideoView *remoteVideoView = (QBRTCRemoteVideoView *)participantView.videoView;
        if (remoteVideoView) {
            [remoteVideoView setVideoTrack:videoTrack];
        }
        participantView.videoContainerView.hidden = NO;
    }
}

- (void)setupConnectionState:(QBRTCConnectionState)connectionState participantId:(NSNumber *)participantId {
    ParticipantView *participantView = (ParticipantView *)[self participantViewWithUserID:participantId];
    if (!participantView) {
        return;
    }
    participantView.connectionState = connectionState;
}

- (void)setupVideoViewHidden:(BOOL)hidden participantId:(NSNumber *)userID {
    ParticipantVideoView *participantView = (ParticipantVideoView *)[self participantViewWithUserID:userID];
    if (!participantView || !participantView.videoView) {
        return;
    }
    participantView.videoContainerView.hidden = hidden;
}

- (void)setupVideoViewAnimation:(CATransition *)animation participantId:(NSNumber *)participantId {
    ParticipantVideoView *participantView = (ParticipantVideoView *)[self participantViewWithUserID:participantId];
    if (!participantView) {
        return;
    }
    UIView *videoView = participantView.videoView;
    if (!videoView) {
        return;
    }
    [videoView.superview.layer addAnimation:animation forKey:nil];
}

- (UIView *)participantViewWithUserID:(NSNumber *)participantId {
    NSArray<ParticipantVideoView *> *participantsViews = [self.topStackView.arrangedSubviews arrayByAddingObjectsFromArray:self.bottomStackView.arrangedSubviews];
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"participantId == %@", participantId];
    return [participantsViews filteredArrayUsingPredicate:predicate].firstObject;;
}

//MARK: - Private Methods
- (UIView *)createParticipantVideoViewWithParticipant:(CallParticipant *)participant {
    ParticipantVideoView *participantView = [[NSBundle mainBundle] loadNibNamed:@"ParticipantVideoView"
                                                                          owner:nil
                                                                        options:nil].firstObject;
    [self setupParticipantView:participantView withParticipant:participant];
    participantView.nameLabel.hidden = participant.id == self.callInfo.localParticipantId;
    if (participant.id != self.callInfo.localParticipantId) {
        QBRTCRemoteVideoView *remoteVideoView = [[QBRTCRemoteVideoView alloc] initWithFrame:CGRectMake(2, 2, 2, 2)];
        remoteVideoView.videoGravity = AVLayerVideoGravityResizeAspect;
        participantView.videoView = remoteVideoView;
    }
    
    return participantView;
}

- (UIView *)createParticipantViewWithParticipant:(CallParticipant *)participant {
    ParticipantView *participantView = [[NSBundle mainBundle] loadNibNamed:@"ParticipantView"
                                                                     owner:nil
                                                                   options:nil].firstObject;
    [self setupParticipantView:participantView withParticipant:participant];
    return participantView;
}

- (void)setupParticipantView:(ParticipantView *)participantView withParticipant:(CallParticipant *)participant {
    participantView.name = participant.fullName;
    participantView.participantId = participant.id;
    if (participant.id == self.callInfo.localParticipantId) {
        participantView.isCallingInfo = NO;
    }
    else if (self.callInfo.direction == CallDirectionIncoming)  {
        participantView.isCallingInfo = NO;
        participantView.stateLabel.text = @"Calling...";
    }
    participantView.connectionState = participant.connectionState;
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Calls/Views/ParticipantsView/ParticipantsView.swift
//
//  ParticipantsView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 16.12.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit
import QuickbloxWebRTC

class ParticipantsView: UIView {
    //MARK: - IBOutlets
    @IBOutlet private weak var topStackView: UIStackView!
    @IBOutlet private weak var bottomStackView: UIStackView!
    //MARK: - Properties
    var callInfo: CallInfo!
    var conferenceType: QBRTCConferenceType = .video
    private var viewCache:[UInt: ParticipantView] = [:]
    
    //MARK: - Public Methods
    func setup(callInfo: CallInfo, conferenceType: QBRTCConferenceType) {
        self.conferenceType = conferenceType
        self.callInfo = callInfo
        
        callInfo.onUpdatedParticipant = { [weak self] (participant) in
            guard let participantView = self?.viewCache[participant.id] else {
                return
            }
            participantView.name = participant.fullName
        }
        
        for participant in callInfo.interlocutors {
            let participantView = conferenceType == .video ? createVideoView(participant) : createView(participant)
            let viewIsFull = topStackView.arrangedSubviews.count == 2
            if viewIsFull {
                bottomStackView.addArrangedSubview(participantView)
                return
            }
            topStackView.addArrangedSubview(participantView)
        }
    }
    
    func addLocalVideo(_ videoView: UIView) {
        if let participantView = viewCache[callInfo.localParticipantId] as? ParticipantVideoView {
            if participantView.videoView != nil {
                return
            }
            participantView.videoView = videoView
            return
        }
        guard let participant = callInfo.participant(callInfo.localParticipantId) else {
            return
        }
        let participantView = createVideoView(participant)
        bottomStackView.addArrangedSubview(participantView)
        participantView.videoView = videoView
    }
    
    func setupVideoTrack(_ videoTrack: QBRTCVideoTrack, participantId: UInt) {
        guard let participantView = viewCache[participantId] as? ParticipantVideoView,
              let participantVideoView = participantView.videoView as? QBRTCRemoteVideoView else {
            return
        }
        participantVideoView.setVideoTrack(videoTrack)
        participantView.videoContainerView.isHidden = false
    }
    
    func hideVideo(_ hidden: Bool, participantId: UInt) {
        guard let participantView = viewCache[participantId] as? ParticipantVideoView else {
            return
        }
        participantView.videoContainerView.isHidden = hidden
    }
    
    func videoAnimation(_ animation: CATransition, participantId: UInt) {
        guard let participantView = viewCache[participantId] as? ParticipantVideoView,
              let videoView = participantView.videoView else {
            return
        }
        videoView.superview?.layer.add(animation, forKey: nil)
    }
    
    func setConnectionState(_ connectionState: QBRTCConnectionState, participantId: UInt) {
        guard let participantView = viewCache[participantId] else {
            return
        }
        participantView.connectionState = connectionState
    }
    
    //MARK: - Private Methods
    private func createVideoView(_ participant: CallParticipant) -> ParticipantVideoView {
        let participantView = ParticipantVideoView.loadNib()
        participantView.nameLabel.isHidden = participant.id == callInfo.localParticipantId
        if participant.id != callInfo.localParticipantId {
            let remoteVideoView = QBRTCRemoteVideoView(frame: CGRect(x: 2.0, y: 2.0, width: 2.0, height: 2.0))
            remoteVideoView.videoGravity = AVLayerVideoGravity.resizeAspect.rawValue
            participantView.videoView = remoteVideoView
        }
        setupView(participantView, participant: participant)
        return participantView
    }
    
    private func createView(_ participant: CallParticipant) -> ParticipantView {
        let participantView = ParticipantView.loadNib()
        setupView(participantView, participant: participant)
        return participantView
    }
    
    private func setupView(_ participantView: ParticipantView, participant: CallParticipant) {
        participantView.name = participant.fullName
        participantView.ID = participant.id
        if participant.id == callInfo.localParticipantId {
            participantView.isCallingInfo = false
        } else if callInfo.direction == .incoming  {
            participantView.isCallingInfo = false
            participantView.stateLabel.text = "Calling..."
        }
        participantView.connectionState = participant.connectionState
        viewCache[participant.id] = participantView
    }
}
